installinator/src/block_size_writer.rs

```diff
@@ -0,0 +1,227 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use pin_project_lite::pin_project;
+use std::io;
+use std::pin::Pin;
+use std::task::ready;
+use std::task::Context;
+use std::task::Poll;
+use tokio::io::AsyncWrite;
+
+pin_project! {
+    /// `BlockSizeWriter` is analogous to a tokio's `BufWriter`, except it
+    /// guarantees that writes made to the underlying writer are always
+    /// _exactly_ the requested block size, with two exceptions: explicitly
+    /// calling (1) `flush()` or (2) `shutdown()` will write any
+    /// buffered-but-not-yet-written data to the underlying buffer regardless of
+    /// its length.
+    ///
+    /// When `BlockSizeWriter` is dropped, any buffered data it's holding will
+    /// be discarded. It is critical to manually call `BlockSizeWriter:flush()`
+    /// or `BlockSizeWriter::shutdown()` prior to dropping to avoid data loss.
+    pub(crate) struct BlockSizeWriter<W> {
```

Comment:
What do you think of calling this `BlockSizeBufWriter`?

Reply:
Sure, seems good to hint at the other related types.

Reply:
Renamed in a5329619